### PR TITLE
Return early from getting state if no checkpoint exists

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Adds defensive checks against nil checkpoints in state and root returning functions.

**Which issues(s) does this PR fix?**

N/A

**Other notes for review**

N/A

### CODE REVIEW

File: beacon-chain/rpc/beaconv1/state.go [hunk 1]
```diff
@@ -231,6 +237,9 @@ func (bs *Server) finalizedStateRoot(ctx context.Context) ([]byte, error) {
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not get finalized checkpoint: %v", err)
 	}
+	if cp == nil {
+		return nil, status.Errorf(codes.NotFound, "Finality not yet achieved")
+	}
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state.go:242]: No test for this code as it's impossible to save a state with empty checkpoint in the DB.

File: beacon-chain/rpc/beaconv1/state.go [hunk 2]
```diff
@@ -246,6 +255,9 @@ func (bs *Server) justifiedStateRoot(ctx context.Context) ([]byte, error) {
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not get justified checkpoint: %v", err)
 	}
+	if cp == nil {
+		return nil, status.Errorf(codes.NotFound, "No justified checkpoint exists")
+	}
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state.go:260]: No test for this code as it's impossible to save a state with empty checkpoint in the DB.

File: beacon-chain/rpc/beaconv1/state.go [hunk 3]
```diff
@@ -174,12 +174,18 @@ func (bs *Server) state(ctx context.Context, stateId []byte) (*statetrie.BeaconS
 		}
 	case "finalized":
 		checkpoint := bs.ChainInfoFetcher.FinalizedCheckpt()
+		if checkpoint == nil {
```
Comment (nisdas) [inline beacon-chain/rpc/beaconv1/state.go:177]: https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/blockchain/chain_info.go#L72

It isn't possible to return a nil checkpoint here.

File: beacon-chain/rpc/beaconv1/state.go [hunk 4]
```diff
@@ -174,12 +174,18 @@ func (bs *Server) state(ctx context.Context, stateId []byte) (*statetrie.BeaconS
...
+			return nil, status.Errorf(codes.NotFound, "Finality not yet achieved")
+		}
 		s, err = bs.StateGenService.StateByRoot(ctx, bytesutil.ToBytes32(checkpoint.Root))
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "Could not get finalized state: %v", err)
 		}
 	case "justified":
 		checkpoint := bs.ChainInfoFetcher.CurrentJustifiedCheckpt()
+		if checkpoint == nil {
```
Comment (nisdas) [inline beacon-chain/rpc/beaconv1/state.go:186]: Same here, it isn't possible to retrieve a nil checkpoint
https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/blockchain/chain_info.go#L81

File: beacon-chain/rpc/beaconv1/state.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/rpc/beaconv1/state.go:242]: So this would be an impossible condition ? Why add a nil check here in that case.

**Review by nisdas — COMMENTED** [review]
All these checks are for impossible conditions, we don't return nil checkpoints anywhere.

**Pull request closed without merging**